CIPs/0010.md

```diff
@@ -0,0 +1,84 @@
+- As part of ODIS, accounts have to authenticate with the ODIS signers for rate-limiting purposes. To do that, the "account key" or the "authorized DEK" has to sign the request which also violates the intention of keeping keys that can move funds as cold as possible.
+- For other applications such as CIP8 or other down-the-road applications, specific keys are going to be useful. Additionally, the `dataEncryptionKey` seems out of place without useful segmentation ability, despite it being effectively also an authorized key (if not signer).
+
+Allowing the segmentation of domains for a Celo account (i.e. an account on `Accounts.sol`) gives Celo stakeholders more flexibility in their custody setup. It allows sensitive domains (like moving balances) to be only accessible by keys that are "cold", i.e. harder to compromise, while making less sensitive domains more user friendly. Expanding such segmentation aids the user experience, both of interacting with the Celo Core Contracts themselves (via authorized smart contracts), as well as dapps that currently would need to copy this authorization scheme to receive similar benefits for their users. While this CIP does not specify further applications, per-dapp-authorizations can become practical.
+
+## Specification
+
+The proposed changes to `Accounts.sol`:
+
+- Add a `signerAuthorizations` struct:
+    ```
+    struct SignerAuthorization {
+      address signer
+      bool authorized
+    }
+    mapping(address => mapping(bytes32 => SignerAuthoriation)) public signerAuthorizations
+    ```
+- Add generalized `authorizeSigner(address signer, string calldata role)` and `completeSignerAuthorization(address account, string calldata role)` for smart contract authorizations (`authorizeSigner` creates a pending authorization that `completeSignerAuthorization` completes)
+- Add a generalized `authorizeSignerWithSignature(address signer, string calldata role, uint8 v, bytes32 r, bytes32 s)` (where v, r, s is an EIP712 signature) for proof-of-possession-based authorization (TODO: EIP712 domain specification)
+
+Both of the methods above would populate the relevant `SignerAuthorization` in the `signerAuthorizations` struct after hashing the `role`, as well as the `authorizedBy` struct. These methods should continue the invariant that a signer is only authorized by a single account ever. However, we can relax the invariant that a signer can only be authorized for at most 1 role.
+
+- Add a generalized `getAuthorizedSigner(address account, string calldata role)` for returning completed authorized signers.
+
+To explicitly retain backwards compatability for existing consumers, the general write methods would special case the roles of, `celo.org/authorizationSchemas/core/vote` and `celo.org/authorizationSchemas/core/attestation`, and also populate the existing `accounts` struct. Similarly, `getAuthorizedSigner` would read from `accounts` in this case.
```

Comment:
Is `authorizationSchemas` included in the path to avoid collision in the hashing domain with other applications?  If so, we might want to drop that and use a constant personalization string prepended to the path when hashing. (Using a personalization string here is a likely a good idea anyway to eliminate a sharp edge for dApp developers who want to use this system.)

Reply:
May you elaborate on this, I'm not sure I'm 100% following

Reply:
Rephrasing: Why is the string `authorizationSchemas` included in the path? It seems clear from the context that this is the usage.

On the personalization string suggestion, it's actually not necessary.  Looking at this a second time I see now that these hashes don't need to be globally unique, as is the case with signatures, so you can ignore that.

Reply:
Gotcha. it's in the string with the possible hope that some kind of structured schema can live at the actual HTTP address eventually. I assume it's at no cost.

Reply:
Should we mention that explicitly here? I imagine more developers will start to use to use this scheme as specified here, and will be choosing their own paths, so it seems important to mention any suggested format for those paths (e.g. that they should point to a unique HTTP URL that they control) so a future application will have a chance to use them in a consistent way.